src/StructuredLogger.Utils/SecretsSearch.cs

```diff
@@ -0,0 +1,140 @@
+            _build = build ?? throw new ArgumentNullException(nameof(build));
+            _search = new Search([build], build.StringTable.Instances, 5000, markResultsInTree: false);
+
+            _detectors = new()
+            {
+                { SensitiveDataKind.CommonSecrets, SensitiveDataDetectorFactory.GetSecretsDetector(SensitiveDataKind.CommonSecrets, false) },
+                { SensitiveDataKind.ExplicitSecrets, SensitiveDataDetectorFactory.GetSecretsDetector(SensitiveDataKind.ExplicitSecrets, false) },
+                { SensitiveDataKind.Username, SensitiveDataDetectorFactory.GetSecretsDetector(SensitiveDataKind.Username, false) }
+            };
+        }
+
+        public bool TryGetResults(NodeQueryMatcher matcher, IList<SearchResult> results, int maxResults)
+        {
+            if (!string.Equals(matcher.TypeKeyword, "secret", StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            var activeDetectors = GetActiveDetectors(matcher.NotMatchers);
+            var foundResults = ScanForSecrets(_build.StringTable.Instances, activeDetectors, maxResults);
+
+            if (foundResults.Any())
+            {
+                foreach (var result in foundResults)
+                {
+                    results.Add(result);
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        public List<SecretDescriptor> SearchSecrets(string text, NodeQueryMatcher? matcher, int maxResults)
+        {
+            var activeDetectors = GetActiveDetectors(matcher == null ? [] : [matcher]);
+
+            if (_secretCache.TryGetValue(text, out var cachedSecrets))
+            {
+                return cachedSecrets.Where(s => activeDetectors.ContainsKey(s.Key)).SelectMany(s => s.Value).Take(maxResults).ToList();
+            }
+
+            var secrets = DetectSecrets(text, activeDetectors);
+            if (secrets.Any())
+            {
+                _secretCache[text] = secrets;
+            }
+
+            return secrets.Values.SelectMany(v => v).Take(maxResults).ToList();
+        }
+
+        private IEnumerable<SearchResult> ScanForSecrets(IEnumerable<string> stringsPool, IReadOnlyDictionary<SensitiveDataKind, ISensitiveDataDetector> detectors, int maxResults)
+        {
+            var results = new List<SearchResult>();
+
+            foreach (var text in stringsPool)
+            {
+                if (_resultCache.TryGetValue(text, out List<SearchResult> cachedResults))
+                {
+                    results.AddRange(cachedResults);
+                    if (results.Count >= maxResults)
+                    {
+                        break;
+                    }
+
+                    continue;
+                }
+
+                var secrets = DetectSecrets(text, detectors);
+
+                if (secrets.Any())
+                {
+                    var nodes = _search.FindNodes(text, default);

```

Comment:
there are two ways to search: this one and the one using SearchIndex.
I think the SearchIndex one should be faster:
https://github.com/KirillOsenkov/MSBuildStructuredLog/blob/fa3c09e998926b1f9cff90ea87456901ff6858d0/src/StructuredLogViewer/Controls/BuildControl.xaml.cs#L90-L106

Reply:
it works better for sure, done